Methods to reflect data changes you've made by other means

christianpbrink opened the pull request:
For use when you want a `PFQueryTableViewController` to reflect a change in your data (that is, a change in the results of its `queryForTable`) without calling `loadObjects` (which you don't want to do if you prefer not to reload the whole table).

christianpbrink committed:
Adds methods to reflect changes in the results of `queryForTable`.  For use when you want to reflect a change in your underlying data without `loadObjects` being called (since that calls `tableView.reloadData` rather than animating the change on a per-row basis).

nlutsenko commented:
Do like the idea of having these methods.
@christianpbrink What do you think if we would rename them to the following:
```objc
- (void)insertObject:(PFObject *)object atIndexPath:(NSIndexPath *)indexPath withRowAnimation:(UITableViewRowAnimation)animation;
- (void)moveObjectFromIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath; // You have this one
- (void)deleteRowsForObjectsAtIndexPaths:(NSArray <NSIndexPath *>*)indexPaths withRowAnimation:(UITableViewRowAnimation)animation;
```

The key points here:
- I don't think you need `willAdd/Move/Delete` methods, since those could be squashed together with the generic one
- Having a control over animation lets people specify directly which animation they want/need
- Delete API could be batch by default, as it's easier to do this way if you batch delete.

Let me know what you think.
Also cc @stanleyw @richardjrossiii

christianpbrink commented:
I like the idea of supplying the animation type.

A concern: these methods should not be named on the same pattern as
`removeObjectsAtIndexPaths:animated:`.  That method is fundamentally
different, since it effects a change in the underlying data rather than
just reflecting the change.

How about `<action>RowForObject` for methods that merely reflect underlying
data changes, and `<action>Object` for methods that actually effect those
changes?  E.g., for the new methods:

- (void)insertRowForObject:(PFObject *)object atIndexPath:(NSIndexPath
*)indexPath withRowAnimation:(UITableViewRowAnimation)animation;
- (void)moveRowForObjectFromIndexPath:(NSIndexPath *)fromIndexPath
toIndexPath:(NSIndexPath *)toIndexPath; // You have this one
- (void)deleteRowsForObjectsAtIndexPaths:(NSArray <NSIndexPath
*>*)indexPaths withRowAnimation:(UITableViewRowAnimation)animation;


On Wed, Nov 25, 2015 at 1:04 AM, Nikita Lutsenko <notifications@github.com>
wrote:

> Do like the idea of having these methods.
> @christianpbrink <https://github.com/christianpbrink> What do you think
> if we would rename them to the following:
>
> - (void)insertObject:(PFObject *)object atIndexPath:(NSIndexPath *)indexPath withRowAnimation:(UITableViewRowAnimation)animation;
> - (void)moveObjectFromIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath; // You have this one
> - (void)deleteRowsForObjectsAtIndexPaths:(NSArray <NSIndexPath *>*)indexPaths withRowAnimation:(UITableViewRowAnimation)animation;
>
> The key points here:
>
>    - I don't think you need willAdd/Move/Delete methods, since those
>    could be squashed together with the generic one
>    - Having a control over animation lets people specify directly which
>    animation they want/need
>    - Delete API could be batch by default, as it's easier to do this way
>    if you batch delete.
>
> Let me know what you think.
> Also cc @stanleyw <https://github.com/stanleyw> @richardjrossiii
> <https://github.com/richardjrossiii>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ParsePlatform/ParseUI-iOS/pull/206#issuecomment-159500095>
> .
>

christianpbrink commented:
Markdown is wacky in that comment, probably because I composed it in gmail.  Sorry.

nlutsenko commented:
Agreed, this is a good approach. The delete one I proposed has that, but I agree that it doesn't hurt to unify all of them. ;)

christianpbrink committed:
Reflecting some changes discussed in https://github.com/ParsePlatform/ParseUI-iOS/pull/206.

I kept `willMoveRowForObjectFromIndexPath:toIndexPath` because I need it for my own project (for the case where `tableView` is in edit mode and the user has just moved a row from one `NSIndexPath` to another manually).

christianpbrink committed:
Fix bug -- `insertRowsForObjects...` was appending instead of inserting at the right indexPaths.

christianpbrink committed:
Merge branch 'develop'

* develop:
  Fix bug -- `insertRowsForObjects...` was appending instead of inserting at the right indexPaths.

nlutsenko commented on ParseUI/Classes/QueryTableViewController/PFQueryTableViewController.h:
```diff
@@ -183,6 +183,28 @@ NS_ASSUME_NONNULL_BEGIN
+/*!
+ @abstract Reflects that there's been a change in the results of `queryForTable`, so that `tableView` as rendered must change -- specifically, `object` must be inserted at `indexPath` -- and `_mutableObjects` must have `object` inserted at `indexPath.row`.
+ */
+- (void)insertRowForObject:(PFObject *)object atIndexPath:(NSIndexPath *)indexPath withRowAnimation:(UITableViewRowAnimation)animation;
```
To simplify the API and remove pollution of namespace, can we remove this one in favor of the batch method at all times? It's super easy to construct arrays if you are adding only a single object, as we have literal syntax. :)

nlutsenko commented on ParseUI/Classes/QueryTableViewController/PFQueryTableViewController.h:
```diff
@@ -183,6 +183,28 @@ NS_ASSUME_NONNULL_BEGIN
+ @abstract Reflects that there's been a change in the results of `queryForTable`, so that `tableView` as rendered must change -- specifically, `object` must be inserted at `indexPath` -- and `_mutableObjects` must have `object` inserted at `indexPath.row`.
+ */
+- (void)insertRowForObject:(PFObject *)object atIndexPath:(NSIndexPath *)indexPath withRowAnimation:(UITableViewRowAnimation)animation;
+- (void)insertRowsForObjects:(NSArray *)objects atIndexPaths:(NSArray *)indexPaths withRowAnimation:(UITableViewRowAnimation)animation;
```
Do you think we could use `NSArray PF_GENERIC(PFObject *)*` and `NSArray PF_GENERIC(NSIndexPath *)*` to specify the expected type of the objects in these arrays?
This applies to all of the new added methods that have `NSArray *` as arguments.

nlutsenko commented:
@christianpbrink As per previous discussion, I would vote to remove the `will*` methods, as I don't think they are super useful at this point, since you can combine both removal from underlying model and removal from table view (which what you do anyway in `insert/move/delete` methods. Another downside that comes to my mind about these methods is the fact that from the naming perspective - it's not quite obvious what are they doing and the fact that these mutate underlying model is an implementation detail.
What do you think?

christianpbrink commented:
Re: the PF_GENERIC thing, I can't make that compile.  Can you show me
exactly what the new method signatures should be?  (Sorry if this is an
Objective-C knowledge thing; I mainly do Swift.)

Re: `will*` methods, I did remove all except the one I need in my own
code.  It's for the scenario where the user has manually moved a row from
one `NSIndexPath` to another in the table view, so that repeating that
change programmatically is incorrect.  I think you have a good point about
the name being unhelpful, though -- any suggestions for a different name?

On Sun, Nov 29, 2015 at 11:22 AM, Nikita Lutsenko <notifications@github.com>
wrote:

> @christianpbrink <https://github.com/christianpbrink> As per previous
> discussion, I would vote to remove the will* methods, as I don't think
> they are super useful at this point, since you can combine both removal
> from underlying model and removal from table view (which what you do anyway
> in insert/move/delete methods. Another downside that comes to my mind
> about these methods is the fact that from the naming perspective - it's not
> quite obvious what are they doing and the fact that these mutate underlying
> model is an implementation detail.
> What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ParsePlatform/ParseUI-iOS/pull/206#issuecomment-160430014>
> .
>

nlutsenko commented:
Sample method signature:
```objc
- (void)deleteRowsForObjectsAtIndexPaths:(NSArray PF_GENERIC(NSIndexPath *)*)indexPaths withRowAnimation:(UITableViewRowAnimation)animation;
```
One thing to note - you need to make sure `#import <Parse/PFConstants.h>` is there in the same header (an import for all of Parse if it exists also works).

As for `will*` methods - what do you think if we would simply implement `UITableViewDataSource` method that reflects the changes for moving rows?
Moving stuff around is very tricky, as the order is not guaranteed to persist, since order is propagated from the server...

cc @richardjrossiii What do you think?
